Use NodesStatsRequestParameters.Metric enums more over string

mhl-b opened the pull request:
Change use of `NodesStatsRequestParameters.Metric` in code from String to Enum.

I found that we use `NodesStatsRequestParameters.Metric` as string and enum interchangeably in code. Enum is already validated at HTTP request processing and can be safely used later in the code. Use of enum makes code simpler and compiler is happy too.

There are many lines touched but these are type changes. Major changes are:
- `NodesStatsRequest` uses `Metric` `Enum` rather than `String`, hence there are no `IllegalStateException` exceptions converting string to enum
- Encapsulate conversions from/to String/Stream into Metric enum. Before it was a bit scattered.

A little cleanup before getting into #110716

mhl-b committed:
use metric enums over string

elasticsearchmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

mhl-b commented on server/src/main/java/org/elasticsearch/rest/action/admin/cluster/RestNodesStatsAction.java:
```diff
@@ -54,17 +54,6 @@ public List<Route> routes() {
         );
     }
 
-    static final Map<String, Consumer<NodesStatsRequest>> METRICS;
-
-    static {
-        Map<String, Consumer<NodesStatsRequest>> map = new HashMap<>();
-        for (NodesStatsRequestParameters.Metric metric : NodesStatsRequestParameters.Metric.values()) {
-            map.put(metric.metricName(), request -> request.addMetric(metric.metricName()));
-        }
-        map.put("indices", request -> request.indices(true));
-        METRICS = Collections.unmodifiableMap(map);
-    }
-
```
moved to Metric

mhl-b commented on server/src/test/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequestTests.java:
```diff
@@ -97,32 +97,6 @@ public void testNodesInfoRequestClear() throws Exception {
         assertThat(request.requestedMetrics(), empty());
     }
 
-    /**
-     * Test that (for now) we can only add metrics from a set of known metrics.
-     */
-    public void testUnknownMetricsRejected() {
-        String unknownMetric1 = "unknown_metric1";
-        String unknownMetric2 = "unknown_metric2";
-        Set<String> unknownMetrics = new HashSet<>();
-        unknownMetrics.add(unknownMetric1);
-        unknownMetrics.addAll(randomSubsetOf(NodesStatsRequestParameters.Metric.allMetrics()));
-
-        NodesStatsRequest request = new NodesStatsRequest();
-
-        IllegalStateException exception = expectThrows(IllegalStateException.class, () -> request.addMetric(unknownMetric1));
-        assertThat(exception.getMessage(), equalTo("Used an illegal metric: " + unknownMetric1));
-
-        exception = expectThrows(IllegalStateException.class, () -> request.removeMetric(unknownMetric1));
-        assertThat(exception.getMessage(), equalTo("Used an illegal metric: " + unknownMetric1));
-
-        exception = expectThrows(IllegalStateException.class, () -> request.addMetrics(unknownMetrics.toArray(String[]::new)));
-        assertThat(exception.getMessage(), equalTo("Used illegal metric: [" + unknownMetric1 + "]"));
-
-        unknownMetrics.add(unknownMetric2);
-        exception = expectThrows(IllegalStateException.class, () -> request.addMetrics(unknownMetrics.toArray(String[]::new)));
-        assertThat(exception.getMessage(), equalTo("Used illegal metrics: [" + unknownMetric1 + ", " + unknownMetric2 + "]"));
-    }
-
```
not need anymore, `addMetric` does not throw

mhl-b committed:
update

idegtiarenko commented on server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequestParameters.java:
```diff
@@ -92,22 +96,41 @@ public enum Metric {
+
+        public static Metric readFrom(StreamInput in) throws IOException {
+            return Metric.get(in.readString());
+        }
```
Just to confirm, are we sure we never serialize unexpected metric?

I am thinking about a scenario when old node serialized unexpected metric name, than then upgraded node will deserialize it as null.

mhl-b replied:
Good point! I missed that. I believe current behaviour is to ignore unknown metrics in transport. I will update code and add test for this.

mhl-b committed:
ignore unknown metrics from transport

mhl-b commented on server/src/test/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequestParametersTests.java:
```diff
@@ -0,0 +1,39 @@
+import java.io.IOException;
+import java.util.Set;
+
+public class NodesStatsRequestParametersTests extends ESTestCase {
+
+    public void testMetricsReadWrite() throws IOException {
+        var out = new BytesRefStreamOutput();
+        var outMetics = Set.of(Metric.OS, Metric.JVM, Metric.ALLOCATIONS);
+        out.writeCollection(outMetics);
+        var in = new ByteArrayStreamInput(out.get().bytes);
+        var inMetrics = Metric.readSetFrom(in);
+        assertEquals(outMetics, inMetrics);
+    }
+
+    public void testUnknownMetricsReadWrite() throws IOException {
+        var out = new BytesRefStreamOutput();
+        var outMetics = Set.of(Metric.OS.metricName(), Metric.JVM.metricName(), "unknown");
+        out.writeStringCollection(outMetics);
+        var in = new ByteArrayStreamInput(out.get().bytes);
+        var inMetrics = Metric.readSetFrom(in);
+        assertEquals("should ignore unknown metrics", Set.of(Metric.OS, Metric.JVM), inMetrics);
+    }
+
+}
```
Added test for reading/writing metrics to stream

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequestParameters.java:
```diff
@@ -17,23 +17,26 @@
 public class NodesStatsRequestParameters implements Writeable {
     private CommonStatsFlags indices = new CommonStatsFlags();
-    private final Set<String> requestedMetrics = new HashSet<>();
+    private Set<Metric> requestedMetrics = new HashSet<>();
```
nit: could we keep this `final`?

mhl-b replied:
Yes, I could.

I think in this case we do more work with keeping it final. When I parse transport message I already allocate a list of strings and a set of metrics. If  `requestedMetrics` is final I need to clear and copy from temporary set, while temporary set is what I need. It's still pretty small set, does not really matter if we copy elements. I dont have preference.

from below:
```
-        requestedMetrics.clear();
-        requestedMetrics.addAll(in.readStringCollectionAsList());
+        requestedMetrics = Metric.readSetFrom(in);
```

idegtiarenko replied:
That is a constructor. I think the same could be achieved with:
```
   private final Set<Metric> requestedMetrics;

    public NodesStatsRequestParameters() {
        requestedMetrics = new HashSet<>();
    }

    public NodesStatsRequestParameters(StreamInput in) throws IOException {
        indices = new CommonStatsFlags(in);
        requestedMetrics = Metric.readSetFrom(in);
...
```

DaveCTurner commented on server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequestParameters.java:
```diff
@@ -92,22 +95,53 @@ public enum Metric {
+            final var names = in.readStringArray();
+            for (var name : names) {
+                if (isValid(name)) {
+                    set.add(metricMap.get(name));
```
Could we instead switch to sending the proper enum values over the wire, and use this bit only for bwc?

mhl-b replied:
I'm not sure how to do that. We do send proper enums within same code version, since we write enums not strings right now. 

For bwc is there trick how to do that? I would assume I need to compare transport version. Do we always increment version when add enum option, how to guarantee that?

idegtiarenko replied:
We can write and read depending on the transport version, something like:
https://github.com/elastic/elasticsearch/blob/3b7b86c507f159180938951e22e611f594325abd/server/src/main/java/org/elasticsearch/cluster/ClusterInfo.java#L116-L120

The version is incremented every time we would like to change a wire format. This is verified by bwc tests or by unit tests if they extend `AbstractWireSerializingTestCase`.

idegtiarenko replied:
>Could we instead switch to sending the proper enum values over the wire, and use this bit only for bwc?

David, could you clarify what do you mean by `proper enum values`? In context of https://github.com/elastic/elasticsearch/pull/107944 I do not think we should rely on `readEnumSet`/`writeEnumSet`.
Did you mean this enum should implement `org.elasticsearch.common.io.stream.Writeable` and we should rely on `org.elasticsearch.common.io.stream.StreamOutput#writeCollection(java.util.Collection<? extends org.elasticsearch.common.io.stream.Writeable>)` ?

DaveCTurner replied:
Yeah I meant `readEnumSet`/`writeEnumSet`. I know they're potentially risky because of BwC but I think we can completely mitigate those risks with a suitable test. I think we should prefer the slimmer wire protocol: even if in this specific case it probably doesn't make much difference, it's the established pattern for sending enum values around.

DaveCTurner reviewed (commented):
:+1: good stuff

mhl-b committed:
Merge remote-tracking branch 'upstream/main' into metrics-enum

mhl-b committed:
change transport version and add bwc test

mhl-b committed:
naming

mhl-b commented:
@DaveCTurner @idegtiarenko 
I addressed feedback. I changed wire format from string to enum and bumped transport version. Added bwc test that covers both formats.

I tried to extend AbstractWireSerializingTestCase, but got into hash and equals test failures. None of these classes and dependencies implemented these methods. It seems it’s good to have, but not required right now, should I add them?

idegtiarenko commented on server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/TransportNodesStatsAction.java:
```diff
@@ -132,26 +131,27 @@ protected NodeStats newNodeResponse(StreamInput in, DiscoveryNode node) throws I
+            metrics.contains(Metric.ADAPTIVE_SELECTION),
+            metrics.contains(Metric.SCRIPT_CACHE),
+            metrics.contains(Metric.INDEXING_PRESSURE),
+            metrics.contains(Metric.REPOSITORIES)
```
This looks much simpler :+1:

idegtiarenko commented on server/src/main/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequestParameters.java:
```diff
@@ -92,22 +96,50 @@ public enum Metric {
-            return metricNames.contains(this.metricName());
+        public static void writeSetTo(StreamOutput out, EnumSet<Metric> metrics) throws IOException {
+            if (out.getTransportVersion().onOrAfter(TransportVersions.USE_NODES_STATS_REQUEST_METRIC_ENUM)) {
+                out.writeEnumSet(metrics);
```
Please do not rely on writeEnumSet/readEnumSet as they are using `Enum:ordinal()` that is easy to break simply my reordering enum values, see https://github.com/elastic/elasticsearch/pull/107944

mhl-b committed:
remove EnumSet from wire

mhl-b committed:
Merge remote-tracking branch 'upstream/main' into metrics-enum

DaveCTurner reviewed (changes requested):
Sorry I'd rather we went back to sending enum values over the wire directly, rather than strings. Let's add a test to make sure we don't inadvertently break BwC rather than establishing a pattern of doing inefficient stuff in production for that purpose.

mhl-b committed:
Merge remote-tracking branch 'upstream/main' into metrics-enum

mhl-b committed:
EnumSet in wire is back

mhl-b commented on server/src/test/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequestParametersTests.java:
```diff
@@ -0,0 +1,73 @@
+        }
+    }
+
+    // future-proof of accidental enum ordering change or extension
+    public void testEnsureMetricOrdinalsOrder() throws IOException {
+        final var ordinalsOrder = new Metric[] {
+            Metric.OS,
+            Metric.PROCESS,
+            Metric.JVM,
+            Metric.THREAD_POOL,
+            Metric.FS,
+            Metric.TRANSPORT,
+            Metric.HTTP,
+            Metric.BREAKER,
+            Metric.SCRIPT,
+            Metric.DISCOVERY,
+            Metric.INGEST,
+            Metric.ADAPTIVE_SELECTION,
+            Metric.SCRIPT_CACHE,
+            Metric.INDEXING_PRESSURE,
+            Metric.REPOSITORIES,
+            Metric.ALLOCATIONS };
+
+        assertArrayEquals("metrics order changed", Metric.values(), ordinalsOrder);
+
+        for (var ordinal = 0; ordinal < ordinalsOrder.length; ordinal++) {
+            var out = new BytesRefStreamOutput();
+            out.writeVInt(ordinal);
+            var in = new ByteArrayStreamInput(out.get().bytes);
+            var metric = in.readEnum(Metric.class);
+            assertEquals(ordinalsOrder[ordinal], metric);
+        }
+    }
```
@DaveCTurner, a test to ensure ordinals are not accidentally changed

mhl-b committed:
naming

mhl-b commented on server/src/test/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequestParametersTests.java:
```diff
@@ -0,0 +1,73 @@
+        }
+    }
+
+    // future-proof of accidental enum ordering change or extension
+    public void testEnsureMetricOrdinalsOrder() throws IOException {
+        final var enumOrder = new Metric[] {
+            Metric.OS,
+            Metric.PROCESS,
+            Metric.JVM,
+            Metric.THREAD_POOL,
+            Metric.FS,
+            Metric.TRANSPORT,
+            Metric.HTTP,
+            Metric.BREAKER,
+            Metric.SCRIPT,
+            Metric.DISCOVERY,
+            Metric.INGEST,
+            Metric.ADAPTIVE_SELECTION,
+            Metric.SCRIPT_CACHE,
+            Metric.INDEXING_PRESSURE,
+            Metric.REPOSITORIES,
+            Metric.ALLOCATIONS };
+
+        assertArrayEquals("metrics order changed", Metric.values(), enumOrder);
+
+        for (var ordinal = 0; ordinal < enumOrder.length; ordinal++) {
+            var out = new BytesRefStreamOutput();
+            out.writeVInt(ordinal);
+            var in = new ByteArrayStreamInput(out.get().bytes);
+            var metric = in.readEnum(Metric.class);
+            assertEquals(enumOrder[ordinal], metric);
+        }
+    }
```
Added enum order test. I dont think writing into wire necessary for ordinals test. Compare arrays should be enough.
```
assertArrayEquals("metrics order changed", Metric.values(), enumOrder);
```

DaveCTurner reviewed (approved):
LGTM2

DaveCTurner commented on server/src/test/java/org/elasticsearch/action/admin/cluster/node/stats/NodesStatsRequestParametersTests.java:
```diff
@@ -0,0 +1,73 @@
+    }
+
+    // future-proof of accidental enum ordering change or extension
+    public void testEnsureMetricOrdinalsOrder() throws IOException {
```
I opened https://github.com/elastic/elasticsearch/pull/111092 because I think we should have a more general-purpose test utility for this sort of thing.

DaveCTurner committed:
Merge remote-tracking branch 'upstream/main' into 0reviews/2024/07/19/mhl-b/metrics-enum

DaveCTurner committed:
Use EnumSerializationTestUtils.assertEnumSerialization

mhl-b committed:
poke ci
